module/userExtension/helper.js

```diff
@@ -211,304 +212,373 @@ module.exports = class UserExtensionHelper {
    * @param {Array} userRolesCSVData
    * @param {Object} userDetails -logged in user details.
    * @param {String} userDetails.id -logged in user id.
+   * @param {Object} tenantAndOrgInfo -tenant and organization information.
    * @returns {Array}
    */
 
-  static bulkCreateOrUpdate(userRolesCSVData, userDetails) {
+  static bulkCreateOrUpdate(userRolesCSVData, userDetails, tenantAndOrgInfo) {
     return new Promise(async (resolve, reject) => {
       try {
+        let userRoleMap = {};
         let userRolesUploadedData = new Array();
+        let aggregateKafkaEventPayloads = [];
+        // Pre-fetch all required data
+        const allProgramIds = new Set();
+        const allUserIds = new Set();
 
-        const userRolesArray = await userRolesHelper.list(
-          {
-            status: 'active',
-            isDeleted: false,
-          },
-          {
-            code: 1,
-            title: 1,
-            entityTypes: 1,
-            isAPlatformRole: 1,
-          },
+        for (const csvRow of userRolesCSVData) {
+          const userRole = gen.utils.valueParser(csvRow);
+
+          if (userRole.programs && userRole.programs.length > 0) {
+            userRole.programs.forEach((programId) => allProgramIds.add(programId));
+          }
+
+          if (userRole.user) {
+            allUserIds.add(userRole.user);
+          }
+        }
+
+        // Fetch program data
+        const allProgramsData = await programsHelper.list(
+          { externalId: { $in: Array.from(allProgramIds) } },
+          ['_id', 'externalId'],
+          '',
+          '',
+          '',
+          tenantAndOrgInfo
         );
 
-        let userRoleMap = {};
-        let userRoleAllowedEntityTypes = {};
-
-        userRolesArray.forEach((userRole) => {
-          userRoleMap[userRole.code] = {
-            roleId: userRole._id,
-            code: userRole.code,
-            entities: [],
-            isAPlatformRole: userRole.isAPlatformRole,
-          };
-          userRoleAllowedEntityTypes[userRole.code] = new Array();
-          if (userRole.entityTypes && userRole.entityTypes.length > 0) {
-            userRole.entityTypes.forEach((entityType) => {
-              userRoleAllowedEntityTypes[userRole.code].push(entityType.entityTypeId);
-            });
+        const programIdMap = {};
```

Comment:
@borkarsaish65 suggesting to provide better comments to the code added. better detailed comments will make things easy for us later

Reply:
added